### PR TITLE
build: switch libdispatch build to cmake

### PR DESCRIPTION
**Pull request by compnerd**
Use the cmake based build system for building libdispatch now that it is
able to build Linux, android, and even possible to cross-compile to
Windows.

Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

### CODE REVIEW

**Comment by modocache** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
swift-corelibs-libdispatch PR#264 should fix the build issue that happened here

**Comment by modocache** [comment]
Please test with following PR:
https://github.com/apple/swift-corelibs-libdispatch/pull/264

@swift-ci Please test

**Comment by compnerd** [comment]
CC @dgrove-oss

**Comment by dgrove-oss** [comment]
I think we aren't ready to merge this PR yet.  

At least for me, I am seeing the libdispatch tests not linking correctly when built with CMake because of missing -L/-rpath to find libSwift.  I will put in a dispatch PR for that today.

@compnerd  have you build the _entire_ source tree (including swift, swift-corelibs-dispatch, swiftpm, swift-corelibs-foundation) from scratch with CMake driving the libdispatch build with this PR in place for build-script?

**Comment by compnerd** [comment]
No, I've not built the entire source tree.  I've been focused on the swift compiler/runtime and libdispatch for now.  I know of at least one other place where we need to tweak things.  But, having an occasional run here is helpful to ensure that I don't accidentally regress the Darwin build.  I've been trying to keep the Darwin build untouched.  Does it not use the autoconf build?  If so, I can greatly simplify this PR.  I know that @das has been working on trying to fix up the CMake build for libdispatch to be sufficient for the macOS build, and I'm really looking forward to that to have a single unified build for libdispatch.

**Comment by compnerd** [comment]
@swift-ci please test

Lets see where we stand

**Comment by gottesmm** [comment]
(Just doing a drive by test since the CI didn't seem to kick in).

**Comment by gottesmm** [comment]
@swift-ci please test

**Comment by gottesmm** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
Interesting, seems that the linker invocation is missing the `-rpath-link` for the swift runtime.

**Comment by dgrove-oss** [comment]
I think the missing -rpath-link is because on the dispatch side it is looking for SWIFT_RUNTIME_LIBDIR to be defined.

File: utils/build-script-impl [hunk 1]
```diff
@@ -2443,41 +2444,60 @@ for host in "${ALL_HOSTS[@]}"; do
...
+                ;;
+                *)
+                  cmake_options=(
+                    ${cmake_options[@]}
+                    -DCMAKE_BUILD_TYPE:STRING="${LIBDISPATCH_BUILD_TYPE}"
+                    -DCMAKE_C_COMPILER:PATH="${LLVM_BIN}/clang"
+                    -DCMAKE_CXX_COMPILER:PATH="${LLVM_BIN}/clang++"
+                    -DCMAKE_SWIFT_COMPILER:PATH="${SWIFTC_BIN}"
+                    -DCMAKE_INSTALL_PREFIX:PATH="$(get_host_install_destdir ${host})$(get_host_install_prefix ${host})"
+                    -DCMAKE_RUNTIME_LIBDIR:PATH="${DISPATCH_TMPLIBDIR}"
```
Comment (dgrove-oss) [inline utils/build-script-impl]: SWIFT_RUNTIME_LIBDIR instead I think

**Comment by compnerd** [comment]
Rebased and adjusted as per @dgrove-oss's point.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
Switched the sourcekit's build to use cmake as well.  Adjusted the build-script to fix the rpath for swiftcore

**Comment by compnerd** [comment]
@swift-ci please test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
Okay, so, the last failure was an install failure.  I think that was caused by me using `${CMAKE_INSTALL_PREFIX}` for the install prefix.  Switch that to `<INSTALL_DIR>` which is the temporary install location for the `ExternalProject`.  Hopefully, this should finally get it right.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
Okay, swift-corelibs-libdispatch PR303 should fix the symlink issue.  It seems that I had them locally which is why I never saw this.  I adjusted the dependencies to deal with BlocksRuntime dependencies being needed as well.

**Comment by compnerd** [comment]
Please test with following pull request:
https://github.com/apple/swift-corelibs-libdispatch/pull/303

@swift-ci Please test Linux platform

**Comment by CodaFi** [comment]
Please test with following pull request:
apple/swift-corelibs-libdispatch#303

@swift-ci Please test Linux platform

**Comment by compnerd** [comment]
@CodaFi thanks!  Great, it now seems to also pass on the build bots!  This is great!  Need to wait for @das to get the PR#303 merged in libdispatch and then we can switch over to the CMake build! ❤️

**Comment by compnerd** [comment]
@swift-ci please smoke test macOS platform

**Comment by compnerd** [comment]
Please test with following pull request:
apple/swift-corelibs-libdispatch#303

@swift-ci Please test macOS platform

**Comment by compnerd** [comment]
Just need to wait for the dispatch PR to get merged before this can be merged.

**Comment by gottesmm** [comment]
@compnerd I am soooo excited about this! Getting rid of autoconf from the build? +1! +1!

**Comment by gottesmm** [comment]
Woot!

**Pull request merged**